<a href="https://colab.research.google.com/github/hongjae6/ML_DL/blob/main/dacon_car_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Baseline EDA에서 한발자국만 더 들어간 데이터 살펴보고 가공해보기


In [1]:
!cp /content/drive/MyDrive/DACON/user_data/sample_submission.csv /content/drive/MyDrive/DACON/user_data/test.csv /content/drive/MyDrive/DACON/user_data/train.csv /content

In [2]:
import pandas as pd

In [3]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')
sample_submission = pd.read_csv('sample_submission.csv')

In [4]:
train.head()

id                          title  odometer location    isimported  \
0   0                   Toyota RAV 4     18277   Lagos   Foreign Used   
1   1            Toyota Land Cruiser        10    Lagos          New    
2   2  Land Rover Range Rover Evoque     83091    Lagos  Foreign Used   
3   3                   Lexus ES 350     91524    Lagos  Foreign Used   
4   4                   Toyota Venza     94177    Lagos  Foreign Used   

           engine transmission    fuel  paint  year    target  
0  4-cylinder(I4)    automatic  petrol    Red  2016  13665000  
1  4-cylinder(I4)    automatic  petrol  Black  2019  33015000  
2  6-cylinder(V6)    automatic  petrol    Red  2012   9915000  
3  4-cylinder(I4)    automatic  petrol   Gray  2007   3815000  
4  6-cylinder(V6)    automatic  petrol    Red  2010   7385000

In [5]:
test.head()

id                title  odometer location    isimported          engine  \
0   0  Mercedes-Benz C 300      1234    Abuja          New   4-cylinder(I4)   
1   1         Honda Accord     29938    Abuja  Foreign Used  4-cylinder(I4)   
2   2  Mercedes-Benz S 550     87501    Lagos  Foreign Used  4-cylinder(I4)   
3   3        Toyota Sienna    180894    Lagos  Locally used  6-cylinder(V6)   
4   4         Toyota Hiace    104814    Lagos  Foreign Used  4-cylinder(I4)   

  transmission    fuel      paint  year  
0    automatic  petrol      White  2017  
1    automatic  petrol      White  2013  
2    automatic  petrol      Black  2012  
3    automatic  petrol  Dark Grey  2001  
4    automatic  petrol      White  2000

title 컬럼부터 살펴보도록 하겠습니다.

In [6]:
sorted(train.title.unique())

['ALPINA B3',
 'Acura MDX',
 'Acura RDX',
 'Acura TL',
 'Acura ZDX',
 'Audi S6',
 'BMW 325',
 'BMW 335',
 'BMW 528',
 'BMW 730',
 'BMW 740',
 'BMW 745',
 'BMW 750',
 'BMW X6',
 'Bentley Arnage',
 'Chevrolet Camaro',
 'Chevrolet Evanda',
 'Chevrolet Traverse',
 'DAF 45',
 'DAF 95XF TRACTOR HEAD',
 'DAF CF',
 'DAF FA 55. 180',
 'Dodge Charger',
 'Ford Edge',
 'Ford Escape',
 'Ford Expedition',
 'Ford Explorer',
 'Ford F 150',
 'Ford Focus',
 'Ford Fusion',
 'Ford Transit',
 'GAC Gonow Other',
 'GMC Terrain',
 'Honda Accord',
 'Honda CR-V',
 'Honda Civic',
 'Honda Crosstour',
 'Honda Element',
 'Honda Fit',
 'Honda Insight',
 'Honda Odyssey',
 'Honda Pilot',
 'Honda Ridgeline',
 'Hyundai Azera',
 'Hyundai Elantra',
 'Hyundai Santa Fe',
 'Hyundai Sonata',
 'Hyundai Tucson',
 'Hyundai ix35',
 'IVECO EUROTECH 7.50E-16',
 'IVM LT35',
 'Infiniti FX',
 'Infiniti JX',
 'Infiniti M35',
 'Infiniti QX80',
 'Isuzu CABSTER',
 'Isuzu NQR',
 'Iveco TRUCK',
 'JMC Vigus',
 'Jaguar X-Type',
 'Jaguar XJ8',

In [7]:
sorted(test.title.unique())

['Acura MDX',
 'Acura RDX',
 'Acura TL',
 'Acura TSX',
 'BMW 525',
 'BMW 535',
 'BMW X3',
 'Chevrolet Camaro',
 'DAF XF TRACTOR HEAD',
 'Fiat Ducato',
 'Ford Edge',
 'Ford Explorer',
 'Ford F 150',
 'Ford Fusion',
 'Ford Mondeo',
 'Ford Mustang',
 'Ford TRANSIT PICKUP',
 'Ford Transit',
 'Honda Accord',
 'Honda CR-V',
 'Honda Civic',
 'Honda Crosstour',
 'Honda Element',
 'Honda Fit',
 'Honda Odyssey',
 'Honda Pilot',
 'Honda Ridgeline',
 'Hyundai Azera',
 'Hyundai Elantra',
 'Hyundai Santa Fe',
 'Hyundai Sonata',
 'Infiniti FX',
 'Iveco EUROCARGO 120e18',
 'Iveco TRUCK',
 'Jeep Wrangler',
 'Kia Optima',
 'Kia Rio',
 'Land Rover Discovery',
 'Land Rover Range Rover',
 'Land Rover Range Rover Sport',
 'Lexus ES 300',
 'Lexus ES 350',
 'Lexus GS 300',
 'Lexus GS 350',
 'Lexus GX 460',
 'Lexus GX 470',
 'Lexus IS 250',
 'Lexus LX 570',
 'Lexus RX 300',
 'Lexus RX 330',
 'Lexus RX 350',
 'Lexus RX 400',
 'MINI Cooper',
 'MINI Cooper Coupé',
 'Mack CXN613 CAB BEHIND ENGINE',
 'Mack R-686ST'

차의 종류가 어마무시하게 많습니다. 들어보지도 못한 차들이 많네요.
많은 차들 중에 train에만 있는 차와 test에만 있는 차도 있을 것 같습니다.


In [8]:
# only_train_data_title = sorted([name for name in train.title.unique() if name not in test.title.unique()])
only_train_data_title = set(train.title.unique()) - set(test.title.unique())
# only_test_data_title = sorted([name for name in test.title.unique() if name not in test.title.unique()])
only_test_data_title = set(test.title.unique()) - set(train.title.unique()) 

In [9]:
print(f'train만 있는 차 종류 : {len(only_train_data_title)}, test만 있는 차 종류 : {len(only_test_data_title)})')

train만 있는 차 종류 : 116, test만 있는 차 종류 : 37)


In [10]:
only_train_data_title

{'ALPINA B3',
 'Acura ZDX',
 'Audi S6',
 'BMW 325',
 'BMW 335',
 'BMW 528',
 'BMW 730',
 'BMW 740',
 'BMW 745',
 'BMW 750',
 'BMW X6',
 'Bentley Arnage',
 'Chevrolet Evanda',
 'Chevrolet Traverse',
 'DAF 45',
 'DAF 95XF TRACTOR HEAD',
 'DAF CF',
 'DAF FA 55. 180',
 'Dodge Charger',
 'Ford Escape',
 'Ford Expedition',
 'Ford Focus',
 'GAC Gonow Other',
 'GMC Terrain',
 'Honda Insight',
 'Hyundai Tucson',
 'Hyundai ix35',
 'IVECO EUROTECH 7.50E-16',
 'IVM LT35',
 'Infiniti JX',
 'Infiniti M35',
 'Infiniti QX80',
 'Isuzu CABSTER',
 'Isuzu NQR',
 'JMC Vigus',
 'Jaguar X-Type',
 'Jaguar XJ8',
 'Jeep Cherokee',
 'Kia Cerato',
 'Kia Picanto',
 'Kia Sorento',
 'Kia Soul',
 'Kia Stinger',
 'Land Rover Range Rover Evoque',
 'Land Rover Range Rover Velar',
 'Lexus CT 200h',
 'Lexus ES 330',
 'Lexus LS 460',
 'Lincoln Mark',
 'MAN-VOLKSWAGEN FLATBED',
 'MINI Cooper Countryman',
 'Mack CH613',
 'Mack CH613 CONVENTIONAL CAB',
 'Mack CXN613',
 'Man Diesel 19.314 FILT-N',
 'Man TGA 18.360',
 'Man Truc

In [11]:
only_test_data_title

{'Acura TSX',
 'BMW 525',
 'BMW 535',
 'BMW X3',
 'DAF XF TRACTOR HEAD',
 'Fiat Ducato',
 'Ford Mondeo',
 'Ford Mustang',
 'Ford TRANSIT PICKUP',
 'Iveco EUROCARGO 120e18',
 'Jeep Wrangler',
 'Land Rover Discovery',
 'Lexus RX 400',
 'MINI Cooper Coupé',
 'Man 26-403',
 'Man LE 8. 180 PLATFORM TRUCK',
 'Mercedes-Benz C 320',
 'Mercedes-Benz CABIN PLUS CHASSIS ONLY',
 'Mercedes-Benz DUMP TRUCK',
 'Mercedes-Benz G 550',
 'Mercedes-Benz GLA 180',
 'Mercedes-Benz GLE 450',
 'Mercedes-Benz GLK 300',
 'Mercedes-Benz Maybach',
 'Mercedes-Benz S 550',
 'Mercedes-Benz WATER TANKER',
 'Mitsubishi Eclipse',
 'Mitsubishi L200',
 'Mitsubishi Montero',
 'Nissan A',
 'Nissan Primera',
 'Peugeot 307',
 'Toyota C-HR',
 'Toyota Prius',
 'Volvo FL12',
 'Volvo FL7',
 'Volvo S80'}

train에만 존재하는 차가 116종류, test에만 존재하는 차가 37종류가 있네요. train의 경우에는 차를 제거하면 되지만, test에 있는 차는 제거할 수 없으니 이를 직접 가공할 방법은 더 생각해봐야겠습니다.

자동차 제조사별로 네이밍 규칙이 다르겠지만, 시리즈가 있는 차들의 경우 (BMW 5 Series) 더 세분화하여 나눠주는 작업도 유의미할 것이라고 생각됩니다. 이름에 제작 모델이 직접 들어가 있지 않은 경우 title과 year를 통해 차량 세대를 알 수 있을 것입니다.


해당 노트북은 Baseline EDA에서 작성된 브랜드를 추출하여 새로운 컬럼을 만드는 방법을 사용하도록 하겠습니다. 같은 등급일수록 브랜드에 따라 가격이 다를 수 있기 때문입니다.


In [12]:
train['title'].apply(lambda x : x.split(" ")[0])

train['brand'] = train['title'].apply(lambda x : x.split(" ")[0])
train.head()

id                          title  odometer location    isimported  \
0   0                   Toyota RAV 4     18277   Lagos   Foreign Used   
1   1            Toyota Land Cruiser        10    Lagos          New    
2   2  Land Rover Range Rover Evoque     83091    Lagos  Foreign Used   
3   3                   Lexus ES 350     91524    Lagos  Foreign Used   
4   4                   Toyota Venza     94177    Lagos  Foreign Used   

           engine transmission    fuel  paint  year    target   brand  
0  4-cylinder(I4)    automatic  petrol    Red  2016  13665000  Toyota  
1  4-cylinder(I4)    automatic  petrol  Black  2019  33015000  Toyota  
2  6-cylinder(V6)    automatic  petrol    Red  2012   9915000    Land  
3  4-cylinder(I4)    automatic  petrol   Gray  2007   3815000   Lexus  
4  6-cylinder(V6)    automatic  petrol    Red  2010   7385000  Toyota

In [13]:
train['brand'].value_counts().head()

Toyota           474
Lexus            122
Mercedes-Benz    105
Honda             81
Ford              36
Name: brand, dtype: int64

In [14]:
test['brand'] = test['title'].apply(lambda x: x.split(" ")[0])
test.head()

id                title  odometer location    isimported          engine  \
0   0  Mercedes-Benz C 300      1234    Abuja          New   4-cylinder(I4)   
1   1         Honda Accord     29938    Abuja  Foreign Used  4-cylinder(I4)   
2   2  Mercedes-Benz S 550     87501    Lagos  Foreign Used  4-cylinder(I4)   
3   3        Toyota Sienna    180894    Lagos  Locally used  6-cylinder(V6)   
4   4         Toyota Hiace    104814    Lagos  Foreign Used  4-cylinder(I4)   

  transmission    fuel      paint  year          brand  
0    automatic  petrol      White  2017  Mercedes-Benz  
1    automatic  petrol      White  2013          Honda  
2    automatic  petrol      Black  2012  Mercedes-Benz  
3    automatic  petrol  Dark Grey  2001         Toyota  
4    automatic  petrol      White  2000         Toyota

이제 location 컬럼을 살펴보도록 하겠습니다.

In [15]:
sorted(train.location.unique())

['Abia',
 'Abia State',
 'Abuja',
 'Abuja ',
 'Accra',
 'Adamawa ',
 'FCT',
 'Lagos',
 'Lagos ',
 'Lagos State',
 'Ogun',
 'Ogun State',
 'other']

In [16]:
sorted(test.location.unique())

['Abia',
 'Abuja',
 'Abuja ',
 'Arepo ogun state ',
 'Lagos',
 'Lagos ',
 'Lagos State',
 'Mushin',
 'Ogun',
 'Ogun State',
 'other']

location의 경우, 뒤의 띄어쓰기가 삽입되어있거나 도시 뒤에 'State'를 붙인  변수가 있습니다. 이를 정리해주겠습니다.

In [17]:
train = train.replace({
    'Abia State' : 'Abia',
    'Abuja ' : 'Abuja',
    'Lagos ' : 'Lagos',
    'Lagos State' : 'Lagos',
    'Ogun State' : 'Ogun'
    })

In [18]:
test = test.replace({
    'Abuja ' : 'Abuja',
    'Lagos ' : 'Lagos',
    'Lagos State' : 'Lagos',
    'Ogun State' : 'Ogun',
    'Arepo ogun state ' : 'Ogun'
    # Arepo is a populated place located in Ogun State, Nigeria. 출처. 위키백과
})

In [19]:
sorted(train.location.unique())

['Abia', 'Abuja', 'Accra', 'Adamawa ', 'FCT', 'Lagos', 'Ogun', 'other']

In [20]:
sorted(test.location.unique())

['Abia', 'Abuja', 'Lagos', 'Mushin', 'Ogun', 'other']

깔끔하게 정리가 되었습니다. Mushin 지역은 train 데이터에 없어 가격을 어떻게 예측할지 모르겠네요. 양쪽에 동시에 있지 않는 값들은 other로 통일해 주는 작업을 하면 더 좋은 성능을 낼 수 있을지 모르겠습니다.

다음 컬럼들을 보겠습니다.

In [21]:
sorted(train.isimported.unique())

['Foreign Used', 'Locally used', 'New ']

In [22]:
sorted(test.isimported.unique())

['Foreign Used', 'Locally used', 'New ']

In [23]:
sorted(train.engine.unique())

['2-cylinder(I2)',
 '3-cylinder(I3)',
 '4-cylinder(H4)',
 '4-cylinder(I4)',
 '5-cylinder(I5)',
 '6-cylinder(I6)',
 '6-cylinder(V6)',
 '8-cylinder(V8)']

In [24]:
sorted(test.engine.unique())

['12-cylinder(V12)',
 '2-cylinder(I2)',
 '3-cylinder(I3)',
 '4-cylinder(I4)',
 '5-cylinder(I5)',
 '6-cylinder(I6)',
 '6-cylinder(V6)',
 '8-cylinder(V8)']

test 데이터셋에서 train 데이터에 없던 '12-cylinder(V12)' 변수가 있습니다. 왠지 실린더가 많으니 가격이 비쌀거 같기도 합니다. test 데이터세트에만 있어 어떻게 처리해야 할 지 조금 더 연구를 해봐야할 것 같습니다.

In [25]:
train.transmission.unique()

array(['automatic', 'manual'], dtype=object)

In [26]:
test.transmission.unique()

array(['automatic', 'manual'], dtype=object)

In [27]:
train.fuel.unique()

array(['petrol', 'diesel'], dtype=object)

In [28]:
test.fuel.unique()

array(['petrol', 'diesel'], dtype=object)

4개의 컬럼들은 손댈 필요가 없는듯 합니다. 그대로 라벨 인코딩을 하면 될 것 같습니다.

색상은 조금 손을 봐야 할 것 같습니다. Baseline EDA에서 제 생각과 같아 코드를 그대로 쓰도록 하겠습니다.

In [29]:
sorted(train.paint.unique())

[' Black',
 ' Black/Red',
 'Ash',
 'Ash and black',
 'BLACK',
 'Beige',
 'Black',
 'Black ',
 'Black and silver',
 'Black sand pearl',
 'Black.',
 'Blue',
 'Blue ',
 'Brown',
 'Cream',
 'Cream ',
 'DARK GREY',
 'Dark Ash',
 'Dark Blue',
 'Dark Green',
 'Dark Grey',
 'Dark ash',
 'Dark blue ',
 'Dark gray',
 'Dark silver ',
 'Deep Blue',
 'Deep blue',
 'GOLD',
 'Gery',
 'Gold',
 'Gold ',
 'Gray',
 'Gray ',
 'Green',
 'Green ',
 'Grey',
 'Grey ',
 'Ink blue',
 'Light Gold',
 'Light blue',
 'Light silver ',
 'Magnetic Gray',
 'Magnetic Gray Metallic',
 'Maroon',
 'Midnight Black Metal',
 'Milk',
 'Navy blue',
 'Off white',
 'Off white l',
 'Pale brown',
 'Purple',
 'Red',
 'Redl',
 'SILVER',
 'Silver',
 'Silver ',
 'Silver/grey',
 'Sky blue',
 'Skye blue',
 'Sliver',
 'Super White',
 'WHITE',
 'WINE',
 'Whine ',
 'White',
 'White ',
 'White orchild pearl',
 'Wine',
 'Yellow',
 'blue',
 'green',
 'orange',
 'red',
 'white',
 'white-blue',
 'yellow']

In [30]:
import re 

def clean_text(texts): 
    corpus = [] 
    for i in range(0, len(texts)): 
        review = re.sub(r'[@%\\*=()/~#&\+á?\xc3\xa1\-\|\.\:\;\!\-\,\_\~\$\'\"\n\]\[\>\<]', '',texts[i]) #@%*=()/+ 와 같은 문장부호 제거
        review = re.sub(r'\d+','',review)#숫자 제거
        review = review.lower() #소문자 변환
        review = re.sub(r'\s+', ' ', review) #extra space 제거
        review = re.sub(r'<[^>]+>','',review) #Html tags 제거
        review = re.sub(r'\s+', ' ', review) #spaces 제거
        review = re.sub(r"^\s+", '', review) #space from start 제거
        review = re.sub(r'\s+$', '', review) #space from the end 제거
        review = re.sub(r'_', ' ', review) #space from the end 제거
        #review = re.sub(r'l', '', review)
        corpus.append(review) 
        
    return corpus

temp = clean_text(train['paint']) #메소드 적용
train['paint'] = temp
print('train data에서 paint의 unique 카테고리 개수 : ', len(train['paint'].unique()))

train data에서 paint의 unique 카테고리 개수 :  51


In [31]:
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'blue' if x.find('blue') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'red' if x.find('red') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'green' if x.find('green') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'white' if x.find('white') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'grey' if x.find('grey') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'grey' if x.find('gery') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'grey' if x.find('gray') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'ash' if x.find('ash') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'brown' if x.find('brown') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'silver' if x.find('silver') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'silver' if x.find('sliver') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'black' if x.find('black') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'gold' if x.find('gold') >= 0 else x)
train['paint'] = train['paint'] = train['paint'].apply(lambda x : 'wine' if x.find('whine') >= 0 else x)

In [32]:
print('train data에서 paint의 unique 카테고리 개수 : ', len(train['paint'].value_counts()))

train data에서 paint의 unique 카테고리 개수 :  18


In [33]:
temp2 = clean_text(test['paint'])
test['paint'] = temp2
print('test data에서 paint의 unique 카테고리 개수 : ', len(test['paint'].unique()))

test data에서 paint의 unique 카테고리 개수 :  30


In [34]:
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'blue' if x.find('blue') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'red' if x.find('red') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'green' if x.find('green') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'white' if x.find('white') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'grey' if x.find('grey') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'grey' if x.find('gery') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'grey' if x.find('gray') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'ash' if x.find('ash') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'brown' if x.find('brown') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'silver' if x.find('silver') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'silver' if x.find('sliver') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'black' if x.find('black') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'gold' if x.find('gold') >= 0 else x)
test['paint'] = test['paint'] = test['paint'].apply(lambda x : 'wine' if x.find('whine') >= 0 else x)

In [35]:
print('test data에서 paint의 unique 카테고리 개수 : ', len(test['paint'].value_counts()))

test data에서 paint의 unique 카테고리 개수 :  18


다음은 year항목을 수정하도록 하겠습니다.

In [36]:
sorted(train.year.unique())

[1217,
 1218,
 1980,
 1982,
 1985,
 1991,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013,
 2014,
 2015,
 2016,
 2017,
 2018,
 2019,
 2020]

In [37]:
sorted(test.year.unique())

[1324,
 1726,
 1987,
 1993,
 1994,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013,
 2014,
 2015,
 2016,
 2017,
 2018,
 2019,
 2020,
 2626]

train 데이터에는 1217년, 1218년 자동차가, test 데이터에는 1324년, 1726년 자동차도 있네요. 저때는 무슨 엔진일까 궁금합니다. 2626년 자동차도 벌써 중고차로 나와있네요. 딜러분께서 영업을 엄청 잘하시는 것 같습니다. 무슨 차인지 보겠습니다.

In [38]:
train[(train.year == 1217) | (train.year == 1218)]

id                title  odometer location    isimported  \
415  415  Mercedes-Benz ATEGO    403461    Lagos  Locally used   
827  827     Mercedes-Benz/52    510053    Lagos  Locally used   

             engine transmission    fuel  paint  year    target  \
415  4-cylinder(I4)       manual  diesel  white  1218   6015000   
827  6-cylinder(V6)       manual  diesel  white  1217  75015000   

                brand  
415     Mercedes-Benz  
827  Mercedes-Benz/52

1218년에 생산된 Mercedes-Benz ATEGO는 트럭이었네요. 1217년에 생산된 Mercedes-Benz/52는 구글에 검색해보니 엔진오일이 가장 비슷한 검색결과로 나옵니다. 두 종류의 차량은 test에 없으니 Baseline 코드를 사용해서 삭제하도록 하겠습니다.

In [39]:
train = train[train['year'] > 1900]

test데이터에도 비정상적인 연도가 있어 살펴보도록 하겠습니다

In [40]:
test[(test.year == 1324) | (test.year == 1726) | (test.year == 2626)]

id                                  title  odometer location  \
304  304  Mercedes-Benz CABIN PLUS CHASSIS ONLY    650923    Lagos   
406  406               Mercedes-Benz DUMP TRUCK         0    Lagos   
411  411             Mercedes-Benz WATER TANKER    701934    Lagos   

       isimported          engine transmission    fuel  paint  year  \
304  Locally used  6-cylinder(V6)       manual  diesel   blue  1324   
406  Locally used  8-cylinder(V8)       manual  diesel  white  2626   
411  Locally used  8-cylinder(V8)       manual  diesel  white  1726   

             brand  
304  Mercedes-Benz  
406  Mercedes-Benz  
411  Mercedes-Benz

train 데이터에 있는 차종이면 그와 비슷하게 맞춰주려고 했는데, test 데이터에만 있는 차종입니다. 304번 차량과 411번 차량의 경우 어느정도 주행거리가 있어 주행거리에 따른 감가상각이 적용되겠지만 406번 차량의 경우는 주행거리마저 0으로 표시되어 있네요. 역시 2626년 차량이라 달린적이 없는 것 같습니다.

304번과 411번 차량의 경우, 주행거리가 정상값이라면, year 변수에 큰 대륙에서 달리는 차량의 평균 주행거리로 계산하여 대입해주면 될 것 같습니다.

406번 차량의 경우 주행거리 0이 정상이라면, year 변수에 최근 모델의 생산연도를 작성하면 어떨까 생각합니다.

꽤나 깊은 고민이 필요한 값일 것 같습니다.

읽어주셔서 감사합니다.
